src/core/eme/session.ts

```diff
@@ -314,174 +304,156 @@ function createSessionAndKeyRequest(
+        const error = new Error("no init data found on media encrypted event.");
+        throw new EncryptedMediaError("INVALID_ENCRYPTED_EVENT", error, true);
       }
 
-      // TODO In that case, the first in $loadedSessions could be this session,
-      // is this wanted?
-      const firstLoadedSession = $loadedSessions.getFirst();
-      if (!firstLoadedSession) {
-        throw error;
-      }
+      const initData = new Uint8Array(encryptedEvent.initData);
+      const initDataType = encryptedEvent.initDataType;
 
-      log.warn("eme: could not create a new session, " +
-               "retry after closing a currently loaded session", error);
+      const generateKeyRequest$ =
+        generateKeyRequest(session, initData, initDataType, "temporary");
 
-      return $loadedSessions.deleteAndClose(firstLoadedSession)
-        .mergeMap(() =>
-          createSessionAndKeyRequest(
-            mediaKeys, keySystem, sessionType, initDataType,
-            initData, errorStream
-          )
-        );
+        return generateKeyRequest$;
     });
 }
 
-/**
- * Create persistent MediaKeySession and load data from a sessionId.
- * @param {MediaKeys} mediaKeys
- * @param {Object} keySystem
- * @param {string} storedSessionId
- * @param {string} initDataType
- * @param {UInt8Array} initData
- * @param {Subject} errorStream
- * @returns {Observable}
- */
-function createPersistentSessionAndLoad(
-  mediaKeys: IMockMediaKeys|MediaKeys,
-  keySystem: IKeySystemOption,
+export function createOrReuseSessionWithRetry(
+  encryptedEvent: MediaEncryptedEvent,
+  mediaKeysInfos: IMediaKeysInfos
+) {
+  return createOrReuseSession(
+    encryptedEvent,
+    mediaKeysInfos
+  ).catch((error) => {
+    if (error.code !== ErrorCodes.KEY_GENERATE_REQUEST_ERROR) {
+      throw error;
+    }
+    const firstLoadedSession = $loadedSessions.getFirst();
+    if (!firstLoadedSession) {
+      throw error;
+    }
+
+    log.warn("eme: could not create a new session, " +
+      "retry after closing a currently loaded session", error);
+
+    return $loadedSessions.deleteAndClose(firstLoadedSession)
+      .mergeMap(() => {
+        return createOrReuseSession(
+          encryptedEvent,
+          mediaKeysInfos
+        );
+      }
+      );
+  });
+}
+
+function createOrReuseSession(
+  encryptedEvent: MediaEncryptedEvent,
+  mediaKeysInfos: IMediaKeysInfos
+): Observable<ISessionEvent> {
+  // reuse currently loaded sessions without making a new key request
+  const initData = new Uint8Array(encryptedEvent.initData || new ArrayBuffer(0));
+  const initDataType = encryptedEvent.initDataType;
+
+  const loadedSession = $loadedSessions.get(initData);
+  if (loadedSession && loadedSession.sessionId) {
+    log.debug("eme: reuse loaded session", loadedSession.sessionId);
+    return Observable.of(createSessionEvent("reuse-session", loadedSession));
+  } else {
+    return createSession(mediaKeysInfos.mediaKeys, "temporary", initData)
```

Comment:
Why always temporary here? Isn't the sessionType wanted in the MKSConfig?

If the guy wants a persistent-license session, we still create a temporary session?

Note: From my understanding two parts are important in a "persistent-licence" config:
  - ``$storedSessions``: initData -> MediaKeySession cache implemented by the implementer (which can make use of the localStorage for example)
  - a MediaKeySession created with the "persistent-licence" sessionType